apis/metal3.io/v1alpha1/hostfirmwaresettings_types.go

```diff
@@ -0,0 +1,81 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+type SchemaReference struct {
+	// `namespace` is the namespace of the where the schema is stored.
+	Namespace string `json:"namespace"`
+	// `name` is the reference to the schema.
+	Name string `json:"name"`
+}
+
+// HostFirmwareSettingsSpec defines the desired state of HostFirmwareSettings
+type HostFirmwareSettingsSpec struct {
+	// FirmwareSchema is a reference to the Schema used to describe each
+	// FirmwareSetting. It can be set here to use a Schema in a different
+	// Namespace other than the one that the settings is in.
+	FirmwareSchema *SchemaReference `json:"schema,omitempty"`
+
+	// Settings are the desired firmware settings stored as name/value pairs.
+	// This will be populated with the actual firmware settings and only
+	// contain the settings that can be modified (i.e. not ReadOnly), to
+	// facilitate making changes.
+	Settings map[string]string `json:"settings" required:"true"`
```

Comment:
If I'm reading https://kubernetes.io/blog/2019/06/20/crd-structural-schema/#extensions correctly, we _don't_ need to set `x-kubernetes-preserve-unknown-fields: true` here because all of the fields are strings. If we had arbitrary JSON data we would need to set that.

It seems like we should consider setting `x-kubernetes-int-or-string: true` so that the user can pass an int in yaml and have it coverted automatically rather than just rejected. It appears the way to trigger that is to use the type k8s.io/apimachinery/pkg/util/intstr.IntOrString in place of string.

Reply:
> It wouldn't hurt to define a type for this like
> 
> ```
> type SettingsMap map[string]string
> ``

Good idea

Reply:
> It seems like we should consider setting `x-kubernetes-int-or-string: true` so that the user can pass an int in yaml and have it coverted automatically rather than just rejected. It appears the way to trigger that is to use the type k8s.io/apimachinery/pkg/util/intstr.IntOrString in place of string.

Yeah using IntOrString seems to work well and would allow the user to set values for example:
"RetryCount": "10"
or
"RetryCount": 10

Thanks for the suggestion.